### DISCUSSION TITLE
removeDuplicates in WithViewStore not working?

### DISCUSSION THREAD

**Discussion by miguelkennedy** [pointfreeco/swift-composable-architecture#495] [Q&A]
Hey guys...

Maybe someone can shed some light into an issue I'm running using SwiftUI and TCA.

I'm implementing removeDuplicates in a WithViewStore initializer in order to avoid re-rendering my view when state doesn't change (the view relies heavily on custom paths and shapes) but I can't seem to make it work. The view still re-renders even though state hasn't change. This ends up affecting the overall performance of my app.

Here's my WithViewStore initializer:

```
WithViewStore(
	self.store.scope(
		state: { $0.chartYAxis },
		action: MainChartFeatureAction.chartYAxis
	),
	removeDuplicates: isDuplicate(lhs:rhs:)
) { viewStore in

REST OF THE VIEW'S CODE GOES HERE

```

And here's how I implemented the isDuplicate function:

```
func isDuplicate(
	lhs: ChartYAxis.ViewState,
	rhs: ChartYAxis.ViewState
) -> Bool {
	return lhs.dummyMode == rhs.dummyMode &&
               lhs.yAxisLineWidth == rhs.yAxisLineWidth &&
	       lhs.scaleYFactor == rhs.scaleYFactor &&
	       lhs.minAxisValue == rhs.minAxisValue &&
	       lhs.maxAxisValue == rhs.maxAxisValue
}
```

By using print() in my code I can see that the isDuplicate function is being called correctly before each render and even though it returns true (state hasn't changed) all views wrapped inside WithViewStore still re-render.

If I redo the entire view using standard SwiftUI code without TCA and using local state and conform the view to Equatable I can avoid re-renders all together as intended.

Can anyone help? What am I doing wrong here? :)

Thanks in advance!

**Comment by SilverTab** [comment]
Was just about to post this _exact_ question but found this one instead, so here's a shameless bump! Even just returning `true` in `removeDuplicates` doesn't seem to have any effect on how often the views inside the `WithViewStore` gets re-rendered, am I perhaps misunderstanding the point of `removeDuplicates`?

Any help would be appreciated!

**Comment by lightandshadow68** [comment]
One way to tell what's causing re-rendering  is to use...

```
var body: some View {
    Self._printChanges()
    return VStack {
        // .. other changes
    }
}
```

Also, you can use `._printChanges()` at the end of a store to print differences as they occur. it's really useful for debugging. 

 `SignInView(store: Store(initialState: SignInFeature.State(), reducer: SignInFeature()._printChanges()))`

See this article for additional details and strategies: https://www.avanderlee.com/swiftui/debugging-swiftui-views/

**Reply by SilverTab** [reply]
Hmm so, after trying that, it looks like no matter what I return in `removeDuplicates:` (wether it's always true, or always false, or some other logic), the output is exactly the same, and `Self._printChanges()` prints the following, the exact same amount of times in each case: `MyView: @self changed.` 🤔 

I was using `_printChanges()` on my store already, so I'm already aware of state changes are occurring, what I don't understand is that I'm purposefully ignoring these state changes in `removeDuplicate` to avoid a re-render, yet it is still happening. Hence why I'm asking if `removeDuplicate` is really what I think it is (i.e. a way to avoid a re-render of the view by returning `true`).

**Reply by stephencelis** [reply]
@SilverTab Without a project to look at it'll be very difficult to debug the issue for you, but if the view is always rendering, then it may be the result of a parent view always updating.

**Reply by SilverTab** [reply]
@stephencelis  Totally understandable! And your comment about the parent view always updating is making me think 🤔 
Will keep that in mind as I take another look at this code, and try to extract a minimal working example to share if possible!

That being said; I'd love to confirm that my understanding of `removeDuplicate`'s purposes is indeed correct! In theory, should returning `true` here prevent the view from getting redrawn?

**Reply by fabstu** [reply]
I found my render instrumentation does a pretty good job of showing why a view re-renders due to TCA. Just applying

```
.tcaSettings(
  .withRerenderLogging(.statistics)
  .withInitialStateHideDiffs(true)
)
```

to the top view shows why ALL the TCA views below re-render.

https://github.com/pointfreeco/swift-composable-architecture/discussions/1695

@stephencelis Are you guys planning to implement something similarly recursive? Currently occasionally rebased to 0.52.11 for my own use, but would love to have something similar inside TCA's debugging-tooling.

**Comment by SilverTab** [comment]
Ok I figured it out! 😅

Turns out @stephencelis was right; the problem was my very root view's ViewStore observing { $0 } which was causing a lot of extra re-renders, so pretty much all the child views where I was trying to narrow the ViewState and the removeDuplicates was getting ignored/bypassed. By reconsidering exactly what the root view needed to observe so as to avoid any unnecessary re-renders, the child views started behaving as I expected!
